### ISSUE TITLE
Conflicting in kind letter F

### ISSUE THREAD

**Issue by masatake** [fishman/ctags#317]
Bad news.

Incompatible change may be needed to fix this issue.

F used as file tag in ctags main part, is also used for other kinds in some parsers.
Here is an example of ruby.

> % cat /tmp/foo.rb 
cat /tmp/foo.rb 
class ClassExample
        def self.class_method
            puts "in class_method"
        end
end        
% ./ctags --extra=+f  /tmp/foo.rb ; cat tags | grep -v '!' | grep F
./ctags --extra=+f  /tmp/foo.rb ; cat tags | grep -v '!' | grep F
class_method	/tmp/foo.rb	/^        def self.class_method$/;"	F	class:ClassExample
foo.rb	/tmp/foo.rb	1;"	F

ctags --list-kinds says:

ObjectiveC
    F  Object field
Ruby
    F  singleton methods
Rust
    F  A method
SQL
    F  record fields

**Comment by masatake** [comment]
About regex parser and xcmd parser,  I would like to forbid to use 'F'.
For crafted build-in parsers I will prepare the way to override 'F' kind letter.

I think Not-changing the meaning of (widely used) tags file is the more important than consistency between parsers. Any comments?

**Comment by vhda** [comment]
I think that would be a nice addition, since "main" uses that specifically to create file tags and we do not control what is added in the regex patterns.
Another option would be to abort when both the language parser has an 'F' kind and the include filenames option is enabled.

On the built in parsers this can be modified, but how can we avoid overlooking updates that add kinds using this letter?
It would be nice if "main" would parse the "kinds" variable and guarantee that no parser would do that, but having such a check at run time is just wasting CPU.
Any other ideas? Maybe using pre-processing code?

**Comment by masatake** [comment]
Thank you for comments.

> I think that would be a nice addition, since "main" uses that specifically to create file tags and we do not control what is added in the regex patterns.
Another option would be to abort when both the language parser has an 'F' kind and the include filenames option is enabled.

If a built-in parser uses 'F' for its own purpose and the parser doesn't define an alternative letter for file, ctags stops the execution(assertion).  This is o.k.? ctags can check in per parser initialization stage. Checking may not be so expensive. Checking is done only once when starting the ctags process; not for each input files.

I would like to prepare incrementally. The initial PR will be for regex parser. tmain can be used forr tesintg.

**Comment by vhda** [comment]
That's fine by me.

**Comment by masatake** [comment]
For the csae that an external command(xcmd) uses 'F' for its own kind, I would like to prepare the
way to override a kind letter for file. The long flag notation I introduced for regex parser can be reused.

   --langdef=MyLang{fileKind=Z}

BTW, Long flag notation for a language may have many applicatons.

  --langdef=SendMailCF{parentParser=m4}

**Comment by masatake** [comment]
About regex parser, I changed my mind.
I wrote code introducing fileKind long flag for xcmd parsers as proposed in the last comment.
The I found the code is applicable to regex parser. The application is straight forward. 
Further more it is not easy to write code permitting overriding file kind in xcmd parsers and forbidding it in regex parsers.  So I will just preapre the way to override the letter for file kind. The long flag for overriding can be used both in xcmd parsers and regex parsers. 

However, when accepting contributions of regex parsers as optlib, we should reject patches using 'F' for its own purpose. In other word using fileKind long flag in regex pasers is for private purpose.

I will withdraw #329 and submit the full set patches.
